src/utils/metric/metric-factory.ts

```diff
@@ -0,0 +1,154 @@
+/*
+	Copyright 2020 Dynatrace LLC
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+import { normalizeDimensions, normalizeMetricKey } from "../normalize";
+import { Dimension, TotalCounter, DeltaCounter, Gauge, Summary, SummaryValue } from "./metric";
+
+export interface Metric {
+	serialize(): string;
+}
+
+export interface MetricFactoryOptions {
+	prefix?: string;
+	defaultDimensions?: Dimension[];
+	oneAgentMetadata?: Dimension[];
+}
+
+export class MetricFactory {
+	private _prefix?: string;
+	private _defaultDimensions: Dimension[];
+	private _oneAgentMetadata: Dimension[];
+
+	/**
+	 * Return a new Metric Factory. If default dimensions are provided, they will be normalized.
+	 */
+	constructor(options?: MetricFactoryOptions) {
+		this._prefix = options?.prefix;
+		this._defaultDimensions = this._deduplicateDimensions(
+			normalizeDimensions(
+				options?.defaultDimensions ?? []
+			)
+		);
+		this._oneAgentMetadata = options?.oneAgentMetadata ?? [];
+	}
+
+	/**
+	 * Create a total counter with a prefixed and normalized key and normalized dimensions.
+	 * Use a total counter when a value represents the total value of a count.
+	 * Returns null if key normalization fails to produce a valid metric key.
+	 */
+	public createTotalCounter(name: string, dimensions: Dimension[], value: number, ts?: number): Metric | null {
```

Comment:
Is `ts` the timestamp? I'd rather spell it out in full and it should also be mentioned in the comment.